librosa.feature.stack_memory returns WRONG VALUES

<!--
If your issue is a usage question, submit it our Google Group instead: 
https://groups.google.com/forum/#!forum/librosa
-->

<!-- Instructions For Filing a Bug: https://github.com/librosa/librosa/blob/master/CONTRIBUTING.md#filing-bugs -->

#### Description
<!-- Example: Error when computing spectral_contrast on a single frame-->
Error when using `librosa.feature.stack_memory ` function, the returned array contains some values that doesn't exists in the input array, and the wrong result is random, the probability of failure increases with bigger `n_steps` parameter

#### Steps/Code to Reproduce
<!--
Example:
```
import librosa

# Generate one frame of random STFT-like energy
S = np.random.randn(1025, 1)**2

contrast = librosa.feature.spectral_contrast(S=S)
```
If the code is too long, feel free to put it in a public gist and link
it in the issue: https://gist.github.com
-->
```python
import librosa
import numpy as np

# generate random array in range [0, 24)
x = np.remainder((np.random.random(10000)*100).astype(int), 24)
assert (x < 24).all(), f'this should not alarm'
stackedX = librosa.feature.stack_memory(
    x, n_steps=500, mode='constant')  # same error with mode type edge and reflect
print(f'{stackedX[stackedX >= 24]}')  # will print very big integers
assert (stackedX < 24).all(), f'this will alarm'

```

#### Expected Results
<!-- Example: contrast should have shape (1, 1)-->
```
$ python test
(nothing happens)
```

#### Actual Results
<!-- Please paste or specifically describe the actual output or traceback. -->
```
$ python test
[2097163]
Traceback (most recent call last):
  File "testST.py", line 10, in <module>
    assert (stackedX < 24).all(), f'this will alarm'
AssertionError: this will alarm
```

#### Versions
<!--
Please run the following snippet and paste the output below.
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import numpy; print("NumPy", numpy.__version__)
import scipy; print("SciPy", scipy.__version__)
import librosa; print("librosa", librosa.__version__)
-->
```python
>>> import platform; print(platform.platform())
Linux-5.0.0-37-generic-x86_64-with-debian-buster-sid
>>> import sys; print("Python", sys.version)
Python 3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0]
>>> import numpy; print("NumPy", numpy.__version__)
NumPy 1.16.4
>>> import scipy; print("SciPy", scipy.__version__)
SciPy 1.3.0
>>> import librosa; print("librosa", librosa.__version__)
librosa 0.7.2
>>> 
```

<!-- Thanks for contributing! -->

Thank you for your reply! I removed the randomness in the following scripts, but i can't use smaller example beacause the code with bug doesn't produce same result every time, and the probability of failure seems to increase with both increased array length and `n_steps`.
```python
import librosa
import numpy as np

# generate random array in range [0, 24)
# x = np.remainder((np.random.random(100000)*100).astype(int), 24)
x = np.remainder(np.arange(100000), 24)
assert (x < 24).all(), f'this should not alarm'
stackedX = librosa.feature.stack_memory(
    x, n_steps=500, mode='constant')  # same error with mode type edge and reflect
print(f'{stackedX[stackedX >= 24]}')  # will print very big integers
assert (stackedX < 24).all(), f'this will alarm'
```

And this is two results:
```
$ python test.py
[2097163 2097175 2097159 2097157 2097158 2097157 2097156 2097154]
Traceback (most recent call last):
  File "testST.py", line 10, in <module>
    print(f'{stackedX[stackedX >= 24]}')  # will print very big integers
AssertionError: this will alarm

$ python test.py
[2097174 2097162 2097165 2097173 2097167 2097155 2097165 2097154 2097152
 2097160 2097167 2097172 2097170 2097162 2097170 2097168]
Traceback (most recent call last):
  File "testST.py", line 11, in <module>
    assert (stackedX < 24).all(), f'this will alarm'
AssertionError: this will alarm
```

i added line 
```python
print(f'x:{x.dtype} sx:{stackedX.dtype}')
```
in the script, the result was:
```
$ python test.py
x:int64 sx:int64
[2097160 2097160 2097161 2097170 2097161 2097155 2097160 2097174 2097157
 2097152 2097168 2097160 2097170 2097161 2097161 2097155 2097161 2097159
 2097163 2097163 2097163 2097175 2097167 2097174 2097171 2097169 2097175
 2097165 2097160 2097161 2097158 2097162 2097171 2097162 2097174 2097157
 2097170 2097156 2097161 2097162 2097158 2097162 2097166 2097159 2097175
 2097160 2097175]
Traceback (most recent call last):
  File "testST.py", line 12, in <module>
    assert (stackedX < 24).all(), f'this will alarm'
AssertionError: this will alarm
```

I tried this:
```python
import librosa
import numpy as np

# generate random array in range [0, 24)
# x = np.remainder((np.random.random(100000)*100).astype(int), 24)
x = np.remainder(np.arange(10000), 24)
assert (x < 24).all(), f'this should not alarm'

sx = np.pad(x, [(500, 0)], mode='constant', constant_values=0)
assert (sx < 24).all()
print('numpy test passed')

stackedX = librosa.feature.stack_memory(
    x, n_steps=500, mode='constant')  # same error with mode type edge and reflect
print(f'x:{x.dtype} sx:{stackedX.dtype}')
print(f'{stackedX[stackedX >= 24]}')  # will print very big integers
assert (stackedX < 24).all(), f'this will alarm'
```
and the bad news is:
```
$ python test.py
numpy test passed
x:int64 sx:int64
[2097159 2097156 2097173 2097162 2097164 2097173 2097162 2097172 2097159
 2097167 2097169 2097171 2097153 2097166 2097163 2097156 2097154 2097154
 2097152 2097172 2097163 2097171 2097175 2097167 2097159 2097152 2097162
 2097170 2097163 2097155 2097162 2097173 2097168 2097160 2097157 2097152
 2097172 2097174 2097166 2097152 2097165 2097154 2097158 2097164 2097174
 2097167 2097173]
Traceback (most recent call last):
  File "testST.py", line 17, in <module>
    assert (stackedX < 24).all(), f'this will alarm'
AssertionError: this will alarm

```

--- a/librosa/feature/utils.py
+++ b/librosa/feature/utils.py
@@ -238,11 +238,7 @@ def stack_memory(data, n_steps=2, delay=1, **kwargs):
 
     data = np.pad(data, [(0, 0), padding], **kwargs)
 
-    history = data
-
-    # TODO: this could be more efficient
-    for i in range(1, n_steps):
-        history = np.vstack([np.roll(data, -i * delay, axis=1), history])
+    history = np.vstack([np.roll(data, -i * delay, axis=1) for i in range(n_steps)[::-1]])
 
     # Trim to original width
     if delay > 0:

--- a/tests/test_features.py
+++ b/tests/test_features.py
@@ -69,9 +69,9 @@ def test_delta_badwidthaxis(x, width, axis):
     librosa.feature.delta(x, width=width, axis=axis)
 
 
-@pytest.mark.parametrize("data", [np.random.randn(5), np.random.randn(5, 5)])
+@pytest.mark.parametrize("data", [np.random.randn(5), np.random.randn(5, 5), np.remainder(np.arange(10000), 24)])
 @pytest.mark.parametrize("delay", [-4, -2, -1, 1, 2, 4])
-@pytest.mark.parametrize("n_steps", [1, 2, 3])
+@pytest.mark.parametrize("n_steps", [1, 2, 3, 300])
 def test_stack_memory(data, n_steps, delay):
 
     data_stack = librosa.feature.stack_memory(data, n_steps=n_steps, delay=delay)
@@ -90,9 +90,11 @@ def test_stack_memory(data, n_steps, delay):
     for i in range(d):
         for step in range(1, n_steps):
             if delay > 0:
-                assert np.allclose(data[i, : -step * delay], data_stack[step * d + i, step * delay :])
+                assert np.allclose(data[i, : -step * delay], data_stack[step * d + i, step * delay:])
             else:
-                assert np.allclose(data[i, -step * delay :], data_stack[step * d + i, : step * delay])
+                assert np.allclose(data[i, -step * delay:], data_stack[step * d + i, : step * delay])
+    assert np.max(data) + 1e-7 >= np.max(data_stack)
+    assert np.min(data) - 1e-7 <= np.min(data_stack)
 
 
 @pytest.mark.parametrize("n_steps,delay", [(0, 1), (-1, 1), (1, 0)])
@@ -299,8 +301,10 @@ def test_rms(y_ex, y2, frame_length, hop_length, center):
     assert y2.size % frame_length == 0
 
     # STFT magnitudes with a constant windowing function and no centering.
-    S1 = librosa.magphase(librosa.stft(y1, n_fft=frame_length, hop_length=hop_length, window=np.ones, center=center))[0]
-    S2 = librosa.magphase(librosa.stft(y2, n_fft=frame_length, hop_length=hop_length, window=np.ones, center=center))[0]
+    S1 = librosa.magphase(librosa.stft(y1, n_fft=frame_length,
+                                       hop_length=hop_length, window=np.ones, center=center))[0]
+    S2 = librosa.magphase(librosa.stft(y2, n_fft=frame_length,
+                                       hop_length=hop_length, window=np.ones, center=center))[0]
 
     # Try both RMS methods.
     rms1 = librosa.feature.rms(S=S1, frame_length=frame_length, hop_length=hop_length)
@@ -347,7 +351,7 @@ def test_zcr_synthetic(y_zcr, frame_length, hop_length, center):
 
     # We don't care too much about the edges if there's padding
     if center:
-        zcr = zcr[:, frame_length // 2 : -frame_length // 2]
+        zcr = zcr[:, frame_length // 2: -frame_length // 2]
 
     # We'll allow 1% relative error
     assert np.allclose(zcr, rate, rtol=1e-2)

